beacon-chain/casper/incentives.go

```diff
@@ -1,37 +1,102 @@
 	"github.com/sirupsen/logrus"
 )
 
 var log = logrus.WithField("prefix", "casper")
 
 // CalculateRewards adjusts validators balances by applying rewards or penalties
 // based on FFG incentive structure.
+// FFG Rewards scheme rewards validator who have voted on blocks, and penalises those validators
+// who are offline. The penalties are more severe the longer they are offline.
 func CalculateRewards(
-	attestations []*pb.AggregatedAttestation,
+	slot uint64,
+	voterIndices []uint32,
 	validators []*pb.ValidatorRecord,
 	dynasty uint64,
-	totalDeposit uint64,
-	totalParticipatedDeposit uint64) ([]*pb.ValidatorRecord, error) {
+	totalParticipatedDeposit uint64,
+	timeSinceFinality uint64) ([]*pb.ValidatorRecord, error) {
+	totalDeposit := TotalActiveValidatorDeposit(dynasty, validators)
 	activeValidators := ActiveValidatorIndices(validators, dynasty)
-	attesterBitfield := attestations[len(attestations)-1].AttesterBitfield
-	attesterFactor := totalParticipatedDeposit * 3
-	totalFactor := totalDeposit * 2
-
-	if attesterFactor >= totalFactor {
-		log.Debug("Applying rewards and penalties for the validators from last cycle")
-		for i, attesterIndex := range activeValidators {
-			voted := shared.CheckBit(attesterBitfield, int(attesterIndex))
-			if voted {
-				validators[i].Balance += params.AttesterReward
-			} else {
-				validators[i].Balance -= params.AttesterReward
+	rewardQuotient := uint64(RewardQuotient(dynasty, validators))
+	PenaltyQuotient := uint64(QuadraticPenaltyQuotient())
+	depositFactor := (totalParticipatedDeposit - totalDeposit) / totalDeposit
+
+	log.Debugf("Applying rewards and penalties for the validators for slot %d", slot)
+	if timeSinceFinality <= (params.CycleLength) {
+		for _, validatorIndice := range activeValidators {
+			var voted bool
+
+			for _, voterIndice := range voterIndices {
+				if voterIndice == validatorIndice {
+					voted = true
+					balance := validators[validatorIndice].GetBalance()
+					newbalance := uint64(balance + (balance/rewardQuotient)*depositFactor)
+					validators[voterIndice].Balance = newbalance
+					break
+				}
+			}
+
+			if !voted {
+				newbalance := validators[validatorIndice].GetBalance()
+				newbalance -= newbalance / rewardQuotient
+				validators[validatorIndice].Balance = newbalance
+			}
+		}
+
+	} else {
+		for _, validatorIndice := range activeValidators {
+			var voted bool
+
+			for _, voterIndice := range voterIndices {
+				if voterIndice == validatorIndice {
+					voted = true
+					break
+				}
+			}
+
+			if !voted {
+				newbalance := validators[validatorIndice].GetBalance()
+				newbalance -= newbalance/rewardQuotient + newbalance*timeSinceFinality/PenaltyQuotient
+				validators[validatorIndice].Balance = newbalance
 			}
 		}
+
 	}
+
 	return validators, nil
 }
+
+// RewardQuotient returns the reward quotient for validators which will be used to
+// reward validators for voting on blocks, or penalise them for being offline.
```

Comment:
RewardQuotient is not used for penalty though? maybe we can take out `, or penalise them for being offline`

Reply:
Reward Quotient is also used for applying penalties for offline validators. The penalty quotient is applied when `timeSinceFinality` > `cycleLength`